pkg/test/test.go

```diff
@@ -0,0 +1,171 @@
+package test
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+	"testing"
+
+	petname "github.com/dustinkirkland/golang-petname"
+	testutils "github.com/kudobuilder/kudo/pkg/test/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var testStageRegex = regexp.MustCompile(`^(\d+)-([^.]+)(.yaml)?$`)
+
+// Test contains all of the test stages and the Kubernetes client and other global configuration
+// for a test.
+type Test struct {
+	Stages []*Stage
+	Name   string
+	Dir    string
+
+	Client          client.Client
+	DiscoveryClient discovery.DiscoveryInterface
+	Logger          testutils.Logger
+}
+
+// DeleteNamespace deletes a namespace in Kubernetes after we are done using it.
+func (t *Test) DeleteNamespace(namespace string) error {
+	t.Logger.Log("Deleting namespace:", namespace)
+	return t.Client.Delete(context.TODO(), &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Namespace",
+		},
+	})
+}
+
+// CreateNamespace creates a namespace in Kubernetes to use for a test.
+func (t *Test) CreateNamespace(namespace string) error {
+	t.Logger.Log("Creating namespace:", namespace)
+	return t.Client.Create(context.TODO(), &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Namespace",
+		},
+	})
+}
+
+// TestFactory creates a new Go test that runs a set of test stages.
+func (t *Test) TestFactory() func(*testing.T) {
+	return func(test *testing.T) {
+		t.Logger = testutils.NewTestLogger(test, t.Name)
+
+		test.Parallel()
+
+		ns := fmt.Sprintf("kudo-test-%s", petname.Generate(2, "-"))
```

Comment:
When running `make integration-test` I am getting:

```
--- FAIL: TestKudoFrameworks (12.13s)
    --- FAIL: TestKudoFrameworks/harness (0.00s)
        --- FAIL: TestKudoFrameworks/harness/framework-test (3.02s)
            logger.go:37: 17:07:06 | framework-test | Creating namespace: kudo-test-complete-silkworm
            logger.go:37: 17:07:06 | framework-test/0-install | starting test stage 0-install
            logger.go:37: 17:07:06 | framework-test/0-install | creating resource: Instance:kudo-test-complete-silkworm/my-instance
            logger.go:37: 17:07:07 | framework-test/0-install | test stage completed 0-install
            logger.go:37: 17:07:07 | framework-test/1-upgrade | starting test stage 1-upgrade
            logger.go:37: 17:07:08 | framework-test/1-upgrade | updating resource: Instance:kudo-test-complete-silkworm/my-instance
            test.go:86: Operation cannot be fulfilled on instances.kudo.k8s.io "my-instance": the object has been modified; please apply your changes to the latest version and try again
            test.go:89: failed in stage 1-upgrade
            logger.go:37: 17:07:09 | framework-test | Deleting namespace: kudo-test-complete-silkworm
```

Reply:
I think we’re seeing a race condition in TestStage’s Create method, KUDO modifies the Instance resource between when we GET the instance and when we UPDATE it. I’ll fix that.

Reply:
This should be fixed now! Whenever you have a chance, could you pull and retry?

Reply:
lgtm